File: nova/mobile/command/entity_extractor.py
"""Entity Extractor Python Binding."""

import re
from typing import Dict
from nova.mobile.command.registry import BuiltInIntent
from nova.mobile.command.context import ExecutionContext


class EntityExtractor:
    def extract(self, text: str, intent: BuiltInIntent, context: ExecutionContext) -> Dict[str, str]:
        entities = {}
        clean = text.strip()

        if intent == BuiltInIntent.CALL_CONTACT:
            name = re.sub(r"(?i)^(call|phone|dial)\s+", "", clean)
            name = re.sub(r"(?i)\s+(on whatsapp|on phone|app)$", "", name).strip()
            resolved = context.resolve_contact(name)
            if resolved:
                entities["contact"] = resolved
                context.update_context(contact=resolved)

        elif intent == BuiltInIntent.SEND_SMS:
            # Match target name after "send sms to" or "send message to" or "send <target> a message"
            match_to = re.search(r"(?i)^(?:send sms to|send message to|text|msg)\s+([a-zA-Z0-9_\s]+?)(?:\s+message|\s+saying|$)", clean)
            match_him = re.search(r"(?i)^send\s+([a-zA-Z0-9_\s]+?)\s+a\s+message", clean)

            target = ""
            if match_to:
                target = match_to.group(1).strip()
            elif match_him:
                target = match_him.group(1).strip()
            else:
                target = re.sub(r"(?i)^(send sms|send message|text|msg)\s*", "", clean).strip()

            target = re.sub(r"(?i)\s+(on whatsapp|on phone|app)$", "", target).strip()
            resolved = context.resolve_contact(target)
            if resolved:
                entities["contact"] = resolved
                context.update_context(contact=resolved)

            msg_match = re.search(r"(?i)saying\s+(.*)", clean)
            if not msg_match:
                msg_match = re.search(r"(?i)message\s+(?:that|saying)?\s*(.*)", clean)
            if msg_match and msg_match.group(1).strip():
                entities["message"] = msg_match.group(1).strip()

        elif intent in (BuiltInIntent.OPEN_APP, BuiltInIntent.PLAY_YOUTUBE, BuiltInIntent.PLAY_SPOTIFY, BuiltInIntent.OPEN_CAMERA, BuiltInIntent.OPEN_BROWSER, BuiltInIntent.OPEN_SETTINGS, BuiltInIntent.OPEN_GALLERY):
            app = re.sub(r"(?i)^(open|launch|start|play|play music|play song)\s+", "", clean)
            app = re.sub(r"(?i)\s+(on phone|app|on youtube|on spotify)$", "", app).strip()
            if app:
                entities["app"] = app
                context.update_context(app=app)

        elif intent in (BuiltInIntent.SET_VOLUME, BuiltInIntent.SET_BRIGHTNESS):
            num_match = re.search(r"(\d{1,3})", clean)
            if num_match:
                entities["percentage"] = num_match.group(1)

        elif intent == BuiltInIntent.SEARCH_GOOGLE:
            query = re.sub(r"(?i)^(search|search google for|google)\s+", "", clean).strip()
            if query:
                entities["query"] = query

        elif intent == BuiltInIntent.OPEN_MAPS:
            dest = re.sub(r"(?i)^(navigate to|maps to|open maps)\s+", "", clean).strip()
            if dest:
                entities["destination"] = dest

        return entities

File: mobile/command/EntityExtractor.kt
package com.nova.mobile.command

import java.util.regex.Pattern

class EntityExtractor {

    fun extract(text: String, intent: BuiltInIntent, context: ExecutionContext): Map<String, String> {
        val entities = mutableMapOf<String, String>()
        val clean = text.trim()

        when (intent) {
            BuiltInIntent.CALL_CONTACT -> {
                val name = parseTargetName(clean, listOf("call", "phone", "dial"))
                val resolved = context.resolveContact(name)
                if (resolved.isNotBlank()) {
                    entities["contact"] = resolved
                    context.updateContext(contact = resolved)
                }
            }

            BuiltInIntent.SEND_SMS -> {
                val name = parseTargetName(clean, listOf("send sms to", "send message to", "text", "msg"))
                val resolved = context.resolveContact(name)
                if (resolved.isNotBlank()) {
                    entities["contact"] = resolved
                    context.updateContext(contact = resolved)
                }
                val msgContent = parseMessageBody(clean)
                if (msgContent.isNotBlank()) {
                    entities["message"] = msgContent
                }
            }

            BuiltInIntent.OPEN_APP -> {
                val app = parseAppName(clean)
                if (app.isNotBlank()) {
                    entities["app"] = app
                    context.updateContext(app = app)
                }
            }

            BuiltInIntent.PLAY_YOUTUBE, BuiltInIntent.PLAY_SPOTIFY -> {
                val query = parseMediaQuery(clean)
                if (query.isNotBlank()) {
                    entities["query"] = query
                }
            }

            BuiltInIntent.SET_VOLUME, BuiltInIntent.SET_BRIGHTNESS -> {
                val percent = parsePercentage(clean)
                if (percent >= 0) {
                    entities["percentage"] = percent.toString()
                }
            }

            BuiltInIntent.SEARCH_GOOGLE -> {
                val query = clean.replace(Regex("(?i)^(search|search google for|google)"), "").trim()
                if (query.isNotBlank()) entities["query"] = query
            }

            BuiltInIntent.OPEN_MAPS -> {
                val dest = clean.replace(Regex("(?i)^(navigate to|maps to|open maps)"), "").trim()
                if (dest.isNotBlank()) entities["destination"] = dest
            }

            else -> {}
        }

        return entities
    }

    private fun parseTargetName(text: String, prefixes: List<String>): String {
        var result = text
        for (prefix in prefixes) {
            val pattern = Pattern.compile("(?i)^" + Pattern.quote(prefix) + "\\s+")
            val matcher = pattern.matcher(result)
            if (matcher.find()) {
                result = matcher.replaceFirst("")
                break
            }
        }
        return result.replace(Regex("(?i)\\s+(on whatsapp|on phone|app)$"), "").trim()
    }

    private fun parseAppName(text: String): String {
        return text.replace(Regex("(?i)^(open|launch|start)\\s+"), "")
            .replace(Regex("(?i)\\s+(on phone|app)$"), "").trim()
    }

    private fun parseMediaQuery(text: String): String {
        return text.replace(Regex("(?i)^(play|play music|play song)\\s+"), "")
            .replace(Regex("(?i)\\s+(on youtube|on spotify|app)$"), "").trim()
    }

    private fun parsePercentage(text: String): Int {
        val p = Pattern.compile("(\\d{1,3})\\s*%?")
        val m = p.matcher(text)
        if (m.find()) {
            val valInt = m.group(1)?.toIntOrNull() ?: -1
            if (valInt in 0..100) return valInt
        }
        return -1
    }

    private fun parseMessageBody(text: String): String {
        val p = Pattern.compile("(?i)message\\s+(?:that|saying)?\\s+(.*)")
        val m = p.matcher(text)
        if (m.find()) {
            return m.group(1)?.trim() ?: ""
        }
        return ""
    }
}
